File: rust/src/day_05.rs
use smtlib::backend::z3_binary::Z3Binary;
use smtlib::prelude::StaticSorted;
use smtlib::{Int, SatResultWithModel, Solver, Storage};
use std::collections::{HashMap, HashSet};

fn parse() -> (Vec<(u32, u32)>, Vec<Vec<u32>>) {
    let content = include_str!("../inputs/input05");
    let mut sections = content.split("\n\n");

    let rules: Vec<(u32, u32)> = sections
        .next()
        .unwrap()
        .lines()
        .map(|line| {
            let (a, b) = line.split_once("|").unwrap();
            (a.parse().unwrap(), b.parse().unwrap())
        })
        .collect();

    let updates: Vec<Vec<u32>> = sections
        .next()
        .unwrap()
        .lines()
        .map(|line| line.split(",").map(|s| s.parse().unwrap()).collect())
        .collect();

    (rules, updates)
}

pub fn solve(part1: bool) -> Result<u32, Box<dyn std::error::Error>> {
    let tup = parse();
    let rules: Vec<(u32, u32)> = tup.0;

    let updates = &tup.1;
    let mut sum: u32 = 0;

    for update in updates {
        let st = Storage::new();
        let mut solver: Solver<Z3Binary> = Solver::new(
            &st,
            Z3Binary::new("z3").expect("Failed to create Z3 binary"),
        )
        .expect("Failed to create solver");

        let mut all_variables: HashSet<u32> = HashSet::new();

        rules.iter().for_each(|(x, y)| {
            if update.contains(&x) && update.contains(&y) {
                let cx = Int::new_const(&st, &format!("v{}", x));
                let cy = Int::new_const(&st, &format!("v{}", y));
                solver.assert(cx.lt(cy)).expect("Oh no!");
                all_variables.insert(*x);
                all_variables.insert(*y);
            }
        });

        for x in &all_variables {
            let var = Int::new_const(&st, &format!("v{}", x));
            solver.assert(var.ge(0)).unwrap(); // var >= 0
        }

        let temp_sum: u32 = match solver.check_sat_with_model().expect("Error while solving") {
            SatResultWithModel::Sat(model) => {
                let mut var_values: HashMap<String, u32> = HashMap::new();
                for x in &all_variables {
                    let var = Int::new_const(&st, &format!("v{}", x));
                    if let Some(value) = model.eval(*&var) {
                        // Parse the value to u32
                        let val_str = value.to_string();
                        if let Ok(val) = val_str.parse::<u32>() {
                            var_values.insert(x.to_string(), val);
                        }
                    }
                }

                let reverse_var_values: HashMap<u32, String> =
                    var_values.iter().map(|(k, v)| (*v, k.clone())).collect();

                let mut l: Vec<u32> = update
                    .iter()
                    .map(|v| *var_values.get(&v.to_string()).unwrap())
                    .collect();

                if l.is_sorted() && part1 {
                    reverse_var_values
                        .get(&l[l.len() / 2])
                        .unwrap()
                        .parse::<u32>()
                        .unwrap()
                } else if !l.is_sorted() && !part1 {
                    l.sort();
                    reverse_var_values
                        .get(&l[l.len() / 2])
                        .unwrap()
                        .parse::<u32>()
                        .unwrap()
                } else {
                    0
                }
            }
            SatResultWithModel::Unsat => 0,
            SatResultWithModel::Unknown => 0,
        };

        sum += temp_sum;
    }

    Ok(sum)
}

File: python/day_05.py
from z3 import *


def parse():
    with open("inputs/input05", "r") as f:
        content = f.read()

    if content == "":
        raise Exception("Empty input, please add inputs/input05 file.")

    sections = content.split("\n\n")

    rules = []
    for line in sections[0].strip().split("\n"):
        a, b = line.split("|")
        rules.append((int(a), int(b)))

    updates = []
    for line in sections[1].strip().split("\n"):
        update = [int(x) for x in line.split(",")]
        updates.append(update)

    return rules, updates


def solve(part1=True):
    rules, updates = parse()
    total_sum = 0

    for update in updates:
        solver = Solver()

        all_variables = set()
        var_map = {}

        for x, y in rules:
            if x in update and y in update:
                if x not in var_map:
                    var_map[x] = Int(f'v{x}')
                if y not in var_map:
                    var_map[y] = Int(f'v{y}')

                # x|y <=> x < y
                solver.add(var_map[x] < var_map[y])
                all_variables.add(x)
                all_variables.add(y)

        for x in all_variables:
            solver.add(var_map[x] >= 0)

        if solver.check() == sat:
            model = solver.model()

            var_values = {}
            for x in all_variables:
                val = model.evaluate(var_map[x])
                var_values[x] = val.as_long()

            reverse_var_values = {v: k for k, v in var_values.items()}

            l = [var_values[v] for v in update]

            is_sorted = all(l[i] <= l[i + 1] for i in range(len(l) - 1))

            if is_sorted and part1:
                # Part 1: sum middle values of already sorted updates
                mid_val = l[len(l) // 2]
                original_num = reverse_var_values[mid_val]
                total_sum += original_num
            elif not is_sorted and not part1:
                # Part 2: sort and sum middle values of incorrectly ordered updates
                l.sort()
                mid_val = l[len(l) // 2]
                original_num = reverse_var_values[mid_val]
                total_sum += original_num

    return total_sum


if __name__ == "__main__":
    print(f"Part 1: {solve(part1=True)}")
    print(f"Part 2: {solve(part1=False)}")